src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java

```diff
@@ -114,7 +125,19 @@ public void rebuildSessionFromFrame(MenuSession menuSession, CaseSearchHelper ca
                     if (step.getId().equals(neededDatum.getDataId())) {
                         if (entityScreen.referencesContainStep(step.getValue())) {
                             currentStep = step.getValue();
+                            processedSteps.add(step);
                             needsFullInit = ++processedStepsCount == steps.size();
+                        } else {
+                            // This block constructs the message to display then throws the exception
+                            List<String> refsList = entityScreen.getReferences().stream()
+                            .map(ref -> EntityScreen.getReturnValueFromSelection(ref, (EntityDatum) neededDatum, entityScreen.getEvalContext()))
+                            .collect(Collectors.toList());
```

Comment:
instead of `.collect(Collectors.toList())` you should be able to just call `.toList()`

Reply:
[8b47055](https://github.com/dimagi/formplayer/pull/1560/commits/8b47055ee35d51dd9e5d5aad89eb30a9c9c60a81)